drivers/docker/driver.go

```diff
@@ -0,0 +1,1682 @@
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	defer h.dloggerPluginClient.Kill()
+
+	c, err := h.client.InspectContainer(h.container.ID)
+	if err != nil {
+		return fmt.Errorf("failed to inspect container state: %v", err)
+	}
+	if c.State.Running && !force {
+		return fmt.Errorf("must call StopTask for the given task before Destroy or set force to true")
+	}
+
+	if err := h.client.StopContainer(h.container.ID, 0); err != nil {
+		h.logger.Warn("failed to stop container during destroy", "error", err)
+	}
+
+	if err := h.dlogger.Stop(); err != nil {
+		h.logger.Error("failed to stop docker logger process during destroy",
+			"error", err, "logger_pid", h.dloggerPluginClient.ReattachConfig().Pid)
+	}
+
+	if err := d.cleanupImage(h); err != nil {
+		h.logger.Error("failed to cleanup image after destroying container",
+			"error", err)
+	}
+
+	return nil
+}
+
+// cleanupImage removes a Docker image. No error is returned if the image
+// doesn't exist or is still in use. Requires the global client to already be
+// initialized.
+func (d *Driver) cleanupImage(handle *taskHandle) error {
+	if !d.config.ImageGC {
+		return nil
+	}
+
+	coordinator, callerID := d.getDockerCoordinator(client, handle.task)
+	coordinator.RemoveImage(handle.container.Image, callerID)
+
+	return nil
+}
+
+func (d *Driver) InspectTask(taskID string) (*drivers.TaskStatus, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	status := &drivers.TaskStatus{
+		ID:          h.task.ID,
+		Name:        h.task.Name,
+		StartedAt:   h.container.State.StartedAt,
+		CompletedAt: h.container.State.FinishedAt,
+		DriverAttributes: map[string]string{
+			"container_id": h.container.ID,
+		},
+		NetworkOverride: h.net,
+		ExitResult:      h.exitResult,
+	}
+
+	status.State = drivers.TaskStateUnknown
+	if h.container.State.Running {
+		status.State = drivers.TaskStateRunning
+	}
+	if h.container.State.Dead {
+		status.State = drivers.TaskStateExited
+	}
+
+	return status, nil
+}
+
+func (d *Driver) TaskStats(taskID string) (*structs.TaskResourceUsage, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	return h.Stats()
+}
+
+func (d *Driver) TaskEvents(ctx context.Context) (<-chan *drivers.TaskEvent, error) {
+	return d.eventer.TaskEvents(ctx)
+}
+
+func (d *Driver) SignalTask(taskID string, signal string) error {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	sig, err := signals.Parse(signal)
+	if err != nil {
+		return fmt.Errorf("failed to parse signal: %v", err)
+	}
+
+	return h.Signal(sig)
+}
+
+func (d *Driver) ExecTask(taskID string, cmd []string, timeout time.Duration) (*drivers.ExecTaskResult, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	if len(cmd) == 0 {
+		return nil, fmt.Errorf("cmd is required, but was empty")
+	}
+
+	ctx, _ := context.WithTimeout(context.Background(), timeout)
```

Comment:
Always defer cancel.
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
```